# 1. Import module

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# 2. Data Loading

In [2]:
minsu = pd.read_csv('정민수_본선작업현황.csv')
minsu.head()

Unnamed: 0                              Unnamed: 1         Unnamed: 2  \
0         NaN  * 본선작업현황 (기간 : 23.11.08 기준으로 90일 이전기간)                NaN   
1         NaN                                   터미널코드                모선명   
2         NaN                                    BPTG     SITC KWANGYANG   
3         NaN                                    BPTG  SAWASDEE SHANGHAI   
4         NaN                                    BPTG       POS SHANGHAI   

  Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6             Unnamed: 7  \
0        NaN        NaN        NaN        NaN                    NaN   
1       모선항차       선석코드       입항연도         선사                 접안예정일시   
2    STKY011          6       2023        SIT  2023-08-10 오후 7:00:00   
3    SWSH008          7       2023        SKR  2023-08-10 오후 7:00:00   
4    SPSO032          7       2023        POL  2023-08-10 오후 7:00:00   

  Unnamed: 8             Unnamed: 9  ... Unnamed: 11 Unnamed: 12 Unnamed: 13  \
0        NaN                    NaN  ...         NaN         NaN         NaN   
1        NaN                 출항예정일시  ...      작업시작일시      양하완료수량      양하잔여수량   
2        NaN  2023-08-11 오전 5:00:00  ...         NaN           0          67   
3        NaN  2023-08-11 오후 7:00:00  ...         NaN           0         678   
4        NaN  2023-08-11 오전 7:00:00  ...         NaN           0         250   

  Unnamed: 14 Unnamed: 15 Unnamed: 16 Unnamed: 17            Unnamed: 18  \
0         NaN         NaN         NaN         NaN                    NaN   
1      양하합계수량      적하완료수량      적하잔여수량      적하합계수량                   생성일자   
2          67           0          60         60   2023-08-08 오후 8:00:20   
3         678           0         600        600   2023-08-08 오후 4:00:16   
4         250           0         250        250   2023-08-08 오후 4:00:16   

             Unnamed: 19 Unnamed: 20  
0                    NaN         NaN  
1                   수정일자        사용여부  
2  2023-08-09 오후 2:00:14         NaN  
3  2023-08-09 오후 2:00:14         NaN  
4  2023-08-09 오후 2:00:14         NaN  

[5 rows x 21 columns]

In [3]:
status = pd.read_csv('본선작업현황.csv')
status.head()

_id terminalCode berthCode terminalShipVoyageNo terminalShipName  \
0  56343          BCT      true              FITT003         EVER FIT   
1  56344          BCT         3              HAJN019   HEUNG A JANICE   
2  56345          BCT         2              HHAU041    HEUNG A ULSAN   
3  56346          BCT         2              MMUK001    MAERSK MAKUTU   
4  56347         BNCT   1(우현접안)              CNAT031  ATLANTIC BRIDGE   

  shippingCode                  etb                  atb                  etd  \
0          EMA  2023-07-29 13:00:00  2023-07-29 13:45:00  2023-07-30 15:00:00   
1          HAS  2023-07-31 03:00:00  2023-07-31 03:20:00  2023-07-31 13:00:00   
2          HAS  2023-07-29 20:00:00  2023-07-29 20:00:00  2023-07-30 08:00:00   
3          MAE  2023-07-31 06:00:00  2023-07-31 06:25:00  2023-07-31 22:00:00   
4          APC  2023-07-30 19:00:00  2023-07-30 19:06:00  2023-07-31 10:00:00   

     atd         commenceTime dischargeCompleted dischargeRemain  \
0  false  2023-07-29 15:16:00                558           false   
1  false  2023-07-31 04:05:00                 57           false   
2  false  2023-07-30 03:40:00                 33           false   
3  false  2023-07-31 07:50:00                847           false   
4  false  2023-07-30 20:36:00                230           false   

  dischargeTotal loadingCompleted loadingRemain loadingTotal registId  \
0            558              822         false          822   SYSTEM   
1             57              136         false          136   SYSTEM   
2             33               10             9           19   SYSTEM   
3            847              531         false          531   SYSTEM   
4            230              247         false          247   SYSTEM   

             registDatetime  
0  2023-08-01T20:47:15.265Z  
1  2023-08-01T20:47:15.304Z  
2  2023-08-01T20:47:15.310Z  
3  2023-08-01T20:47:15.315Z  
4  2023-08-01T20:47:15.321Z

# 3. Data Preprocessing

## 3-1 ) minsu

In [4]:
# 불필요한 행/열 삭제
minsu = minsu.drop([0, 1]).drop(columns=['Unnamed: 0']).reset_index(drop=True)

In [5]:
# 칼럼명 변경
new_columns = ['터미널코드', '모선명', '터미널모선항차', '선석코드', '입항연도','선사코드','접안예정일시','_1_','출항예정일시','_2_','작업시작일시','양하완료수량','양하잔여수량','양하합계수량','적하완료수량','적하잔여수량','적하합계수량','생성일자', '수정일자', '사용여부']
minsu.columns = new_columns

In [6]:
# 필요한 칼럼만 남기기
col = ['터미널코드', '선석코드', '터미널모선항차', '접안예정일시', '출항예정일시', '양하완료수량', '양하잔여수량', '양하합계수량', '적하완료수량', '적하잔여수량', '적하합계수량',]
minsu = minsu[col]

In [7]:
minsu.head()

터미널코드 선석코드  터미널모선항차                 접안예정일시                 출항예정일시 양하완료수량  \
0  BPTG    6  STKY011  2023-08-10 오후 7:00:00  2023-08-11 오전 5:00:00      0   
1  BPTG    7  SWSH008  2023-08-10 오후 7:00:00  2023-08-11 오후 7:00:00      0   
2  BPTG    7  SPSO032  2023-08-10 오후 7:00:00  2023-08-11 오전 7:00:00      0   
3  BPTG    8  SPSO032  2023-08-10 오후 7:00:00  2023-08-11 오전 7:00:00      0   
4  BPTG    9  HSFR008  2023-08-10 오후 7:00:00  2023-08-11 오전 9:00:00      0   

  양하잔여수량 양하합계수량 적하완료수량 적하잔여수량 적하합계수량  
0     67     67      0     60    60   
1    678    678      0    600   600   
2    250    250      0    250   250   
3    250    250      0    250   250   
4    250    250      0    250   250

                                                -

In [8]:
# 데이터 복사
minsu1 = minsu.copy()

In [9]:
# 데이터 분리 (날짜 / 시각)
minsu1[['접안예정일시_날짜', '접안예정일시_시각']] = minsu1['접안예정일시'].str.split(' ', n=1, expand=True)
minsu1.drop(columns=['접안예정일시'], inplace=True)

minsu1[['출항예정일시_날짜', '출항예정일시_시각']] = minsu1['출항예정일시'].str.split(' ', n=1, expand=True)
minsu1.drop(columns=['출항예정일시'], inplace=True)

In [10]:
# '~_시각' 전처리 함수

def convert_to_24hr(time_str):
    # 데이터가 문자열이 아니거나 결측치인 경우 그대로 반환
    if not isinstance(time_str, str) or pd.isna(time_str):
        return time_str
    
    # 시간 문자열을 공백을 기준으로 분할
    time_parts = time_str.split()
    
    # 분할된 문자열이 유효한지 확인
    if len(time_parts) != 2:
        return time_str
    
    # 오전/오후와 시간을 분리합니다.
    am_pm = time_parts[0].strip()
    time_digits = time_parts[1].strip().split(':')
    
    # 시간과 분을 추출하고 필요에 따라 0을 추가합니다.
    hour = time_digits[0].zfill(2)
    minute = time_digits[1].zfill(2)
    second = time_digits[2].zfill(2) if len(time_digits) > 2 else '00'

    # 12시간제를 24시간제로 변환합니다.
    if am_pm == '오후':
        hour = str(int(hour) + 12)
        if hour == '24':
            hour = '12'
    elif am_pm == '오전' and hour == '12':
        hour = '00'

    return f"{hour}:{minute}:{second}"

In [11]:
# '~시각' 결측치 '00:00:00' 대체
minsu1['접안예정일시_시각'] = minsu1['접안예정일시_시각'].apply(convert_to_24hr).fillna('00:00:00')
minsu1['출항예정일시_시각'] = minsu1['출항예정일시_시각'].apply(convert_to_24hr).fillna('00:00:00')

In [12]:
# '~_날짜' 결측치 '0000_00_00' 대체
minsu1['접안예정일시_날짜'] = minsu1['접안예정일시_날짜'].fillna('0000_00_00')
minsu1['출항예정일시_날짜'] = minsu1['출항예정일시_날짜'].fillna('0000_00_00')

In [13]:
# '~수량' 결측치 0으로 대체

minsu1['양하완료수량'].fillna(0, inplace=True)
minsu1['양하잔여수량'].fillna(0, inplace=True)
minsu1['양하합계수량'].fillna(0, inplace=True)
minsu1['적하완료수량'].fillna(0, inplace=True)
minsu1['적하잔여수량'].fillna(0, inplace=True)
minsu1['적하합계수량'].fillna(0, inplace=True)

In [14]:
# 최종 결측치 0개
minsu1.isna().sum()

터미널코드        0
선석코드         0
터미널모선항차      0
양하완료수량       0
양하잔여수량       0
양하합계수량       0
적하완료수량       0
적하잔여수량       0
적하합계수량       0
접안예정일시_날짜    0
접안예정일시_시각    0
출항예정일시_날짜    0
출항예정일시_시각    0
dtype: int64

In [15]:
minsu1.head()

터미널코드 선석코드  터미널모선항차 양하완료수량 양하잔여수량 양하합계수량 적하완료수량 적하잔여수량 적하합계수량   접안예정일시_날짜  \
0  BPTG    6  STKY011      0     67     67      0     60    60   2023-08-10   
1  BPTG    7  SWSH008      0    678    678      0    600   600   2023-08-10   
2  BPTG    7  SPSO032      0    250    250      0    250   250   2023-08-10   
3  BPTG    8  SPSO032      0    250    250      0    250   250   2023-08-10   
4  BPTG    9  HSFR008      0    250    250      0    250   250   2023-08-10   

  접안예정일시_시각   출항예정일시_날짜 출항예정일시_시각  
0  19:00:00  2023-08-11  05:00:00  
1  19:00:00  2023-08-11  19:00:00  
2  19:00:00  2023-08-11  07:00:00  
3  19:00:00  2023-08-11  07:00:00  
4  19:00:00  2023-08-11  09:00:00

---

## 3-2 ) status

In [16]:
status = pd.read_csv('본선작업현황.csv')
status.head()

_id terminalCode berthCode terminalShipVoyageNo terminalShipName  \
0  56343          BCT      true              FITT003         EVER FIT   
1  56344          BCT         3              HAJN019   HEUNG A JANICE   
2  56345          BCT         2              HHAU041    HEUNG A ULSAN   
3  56346          BCT         2              MMUK001    MAERSK MAKUTU   
4  56347         BNCT   1(우현접안)              CNAT031  ATLANTIC BRIDGE   

  shippingCode                  etb                  atb                  etd  \
0          EMA  2023-07-29 13:00:00  2023-07-29 13:45:00  2023-07-30 15:00:00   
1          HAS  2023-07-31 03:00:00  2023-07-31 03:20:00  2023-07-31 13:00:00   
2          HAS  2023-07-29 20:00:00  2023-07-29 20:00:00  2023-07-30 08:00:00   
3          MAE  2023-07-31 06:00:00  2023-07-31 06:25:00  2023-07-31 22:00:00   
4          APC  2023-07-30 19:00:00  2023-07-30 19:06:00  2023-07-31 10:00:00   

     atd         commenceTime dischargeCompleted dischargeRemain  \
0  false  2023-07-29 15:16:00                558           false   
1  false  2023-07-31 04:05:00                 57           false   
2  false  2023-07-30 03:40:00                 33           false   
3  false  2023-07-31 07:50:00                847           false   
4  false  2023-07-30 20:36:00                230           false   

  dischargeTotal loadingCompleted loadingRemain loadingTotal registId  \
0            558              822         false          822   SYSTEM   
1             57              136         false          136   SYSTEM   
2             33               10             9           19   SYSTEM   
3            847              531         false          531   SYSTEM   
4            230              247         false          247   SYSTEM   

             registDatetime  
0  2023-08-01T20:47:15.265Z  
1  2023-08-01T20:47:15.304Z  
2  2023-08-01T20:47:15.310Z  
3  2023-08-01T20:47:15.315Z  
4  2023-08-01T20:47:15.321Z

In [17]:
# 칼럼명 변경
new_columns2 = ['id','터미널코드','선석코드','터미널모선항차','모선명','선사코드','접안예정일시','모선접안일','출항예정일시','출항일시','작업시작일시','양하완료수량','양하잔여수량','양하합계수량','적하완료수량','적하잔여수량','적하합계수량','등록자','등록시간']
status.columns = new_columns2

In [18]:
# 필요한 칼럼만 남기기
col_ = ['터미널코드', '선석코드', '터미널모선항차', '접안예정일시', '출항예정일시', '양하완료수량', '양하잔여수량', '양하합계수량', '적하완료수량', '적하잔여수량', '적하합계수량',]
status = status[col_]

In [19]:
status.head()

터미널코드     선석코드  터미널모선항차               접안예정일시               출항예정일시 양하완료수량  \
0   BCT     true  FITT003  2023-07-29 13:00:00  2023-07-30 15:00:00    558   
1   BCT        3  HAJN019  2023-07-31 03:00:00  2023-07-31 13:00:00     57   
2   BCT        2  HHAU041  2023-07-29 20:00:00  2023-07-30 08:00:00     33   
3   BCT        2  MMUK001  2023-07-31 06:00:00  2023-07-31 22:00:00    847   
4  BNCT  1(우현접안)  CNAT031  2023-07-30 19:00:00  2023-07-31 10:00:00    230   

  양하잔여수량 양하합계수량 적하완료수량 적하잔여수량 적하합계수량  
0  false    558    822  false    822  
1  false     57    136  false    136  
2  false     33     10      9     19  
3  false    847    531  false    531  
4  false    230    247  false    247

                                                -

In [20]:
# 데이터 복사
status1 = status.copy()

In [21]:
## 'true'값 1로 변환 (minsu 데이터의 '1' 비율 반영)
status1['선석코드'].replace('true', 1, inplace=True)
status1['선석코드'].value_counts()

선석코드
1          618
3          587
2          518
7          316
6          293
4          255
T2         253
T1         242
T3         223
9          205
B4         185
5          180
8          178
B5         157
4B         157
1B         152
4A         138
B6         131
2B         121
1(우현접안)    120
B7         116
E3         112
5A         110
E2         105
E1          98
B8          96
4(우현접안)     89
M2          89
3B          87
2(우현접안)     68
B3          54
B9          43
1A          33
2(좌현접안)     28
M3          27
3(우현접안)     24
3A          20
1(좌현접안)     18
M1          14
2A          12
4(좌현접안)      9
3(좌현접안)      7
Name: count, dtype: int64

In [22]:
# 데이터 분리 (날짜 / 시각)
status1[['접안예정일시_날짜', '접안예정일시_시각']] = status1['접안예정일시'].str.split(' ', n=1, expand=True)
status1.drop(columns=['접안예정일시'], inplace=True)

status1[['출항예정일시_날짜', '출항예정일시_시각']] = status1['출항예정일시'].str.split(' ', n=1, expand=True)
status1.drop(columns=['출항예정일시'], inplace=True)

In [23]:
# '~시각' 결측치 '00:00:00' 대체
status1['접안예정일시_시각'] = status1['접안예정일시_시각'].apply(convert_to_24hr).fillna('00:00:00')
status1['출항예정일시_시각'] = status1['출항예정일시_시각'].apply(convert_to_24hr).fillna('00:00:00')

In [24]:
# '~_날짜' 결측치 '0000_00_00' 대체
status1['접안예정일시_날짜'] = status1['접안예정일시_날짜'].fillna('0000_00_00')
status1['출항예정일시_날짜'] = status1['출항예정일시_날짜'].fillna('0000_00_00')

In [25]:
# '~수량' 결측치 0으로 대체
status1['양하완료수량'].replace('false', 0, inplace=True)
status1['양하잔여수량'].replace('false', 0, inplace=True)
status1['양하합계수량'].replace('false', 0, inplace=True)
status1['적하완료수량'].replace('false', 0, inplace=True)
status1['적하잔여수량'].replace('false', 0, inplace=True)
status1['적하합계수량'].replace('false', 0, inplace=True)

In [26]:
# 최종 'false'값 0개
false_cnt = (status1 == 'false').sum().sum()
false_cnt

0

In [27]:
status1.head()

터미널코드     선석코드  터미널모선항차 양하완료수량 양하잔여수량 양하합계수량 적하완료수량 적하잔여수량 적하합계수량  \
0   BCT        1  FITT003    558      0    558    822      0    822   
1   BCT        3  HAJN019     57      0     57    136      0    136   
2   BCT        2  HHAU041     33      0     33     10      9     19   
3   BCT        2  MMUK001    847      0    847    531      0    531   
4  BNCT  1(우현접안)  CNAT031    230      0    230    247      0    247   

    접안예정일시_날짜 접안예정일시_시각   출항예정일시_날짜 출항예정일시_시각  
0  2023-07-29  13:00:00  2023-07-30  15:00:00  
1  2023-07-31  03:00:00  2023-07-31  13:00:00  
2  2023-07-29  20:00:00  2023-07-30  08:00:00  
3  2023-07-31  06:00:00  2023-07-31  22:00:00  
4  2023-07-30  19:00:00  2023-07-31  10:00:00

---

## 4. Data Comparison

In [28]:
minsu1.head()

터미널코드 선석코드  터미널모선항차 양하완료수량 양하잔여수량 양하합계수량 적하완료수량 적하잔여수량 적하합계수량   접안예정일시_날짜  \
0  BPTG    6  STKY011      0     67     67      0     60    60   2023-08-10   
1  BPTG    7  SWSH008      0    678    678      0    600   600   2023-08-10   
2  BPTG    7  SPSO032      0    250    250      0    250   250   2023-08-10   
3  BPTG    8  SPSO032      0    250    250      0    250   250   2023-08-10   
4  BPTG    9  HSFR008      0    250    250      0    250   250   2023-08-10   

  접안예정일시_시각   출항예정일시_날짜 출항예정일시_시각  
0  19:00:00  2023-08-11  05:00:00  
1  19:00:00  2023-08-11  19:00:00  
2  19:00:00  2023-08-11  07:00:00  
3  19:00:00  2023-08-11  07:00:00  
4  19:00:00  2023-08-11  09:00:00

In [29]:
status1.head()

터미널코드     선석코드  터미널모선항차 양하완료수량 양하잔여수량 양하합계수량 적하완료수량 적하잔여수량 적하합계수량  \
0   BCT        1  FITT003    558      0    558    822      0    822   
1   BCT        3  HAJN019     57      0     57    136      0    136   
2   BCT        2  HHAU041     33      0     33     10      9     19   
3   BCT        2  MMUK001    847      0    847    531      0    531   
4  BNCT  1(우현접안)  CNAT031    230      0    230    247      0    247   

    접안예정일시_날짜 접안예정일시_시각   출항예정일시_날짜 출항예정일시_시각  
0  2023-07-29  13:00:00  2023-07-30  15:00:00  
1  2023-07-31  03:00:00  2023-07-31  13:00:00  
2  2023-07-29  20:00:00  2023-07-30  08:00:00  
3  2023-07-31  06:00:00  2023-07-31  22:00:00  
4  2023-07-30  19:00:00  2023-07-31  10:00:00

In [30]:
# 인덱스 재설정
status1.reset_index(drop=True, inplace=True)
minsu1.reset_index(drop=True, inplace=True)

# 두 데이터프레임의 열 순서를 맞춤
minsu1 = minsu1[status1.columns]

# 데이터프레임을 머지하여 일치하는 행 확인
merged_rows = pd.merge(minsu1, status1, how='inner', left_index=True, right_index=True).shape[0]

# 전체 행 수 파악
minsu_rows = minsu1.shape[0]
status_rows = status1.shape[0]


# 일치하는 비율 계산
ratio_m = merged_rows / minsu_rows
ratio_s = merged_rows / status_rows

print(merged_rows)
print(minsu_rows)
print(status_rows)
print("minsu 데이터에 대한 일치비율:", ratio_m)
print("status 데이터에 대한 일치비율:", ratio_s)

4158
4158
6288
minsu 데이터에 대한 일치비율: 1.0
status 데이터에 대한 일치비율: 0.6612595419847328


---

## 5. Conclusion

> #### 두 데이터의 일치 정도: 66%
( 단, 데이터 수집 기간이 다름을 감안해야 함. )